table.go

```diff
@@ -1391,3 +1072,85 @@ func (t *Table) close() {
 	t.active.pendingWritersWg.Wait()
 	t.closing = true
 }
+
+// configureLSMLevels configures the level configs for this table.
+func (t *Table) configureLSMLevels(levels []*LevelConfig) []*LevelConfig {
+	config := make([]*LevelConfig, 0, len(levels))
+
+	for _, level := range levels {
+		config = append(config, &LevelConfig{
+			Level:   level.Level,
+			MaxSize: level.MaxSize,
+			Compact: t.parquetCompaction, // NOTE: right now we only support parquet compaction but in the future this could compact parts into larger arrow records.
+		})
+	}
+
+	// Don't set the compaction functino for the last level. As the LSM doesn't have a level after the last one.
+	config[len(levels)-1].Compact = nil
+
+	return config
+}
+
+func (t *Table) parquetCompaction(compact []*parts.Part) (*parts.Part, int64, int64, error) {
+	b := &bytes.Buffer{}
+	size, err := t.compactParts(b, compact)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	buf, err := dynparquet.ReaderFromBytes(b.Bytes())
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	return parts.NewPart(0, buf), size, int64(b.Len()), nil
+}
+
+func (t *Table) externalParquetCompaction(writer io.Writer) func(compact []*parts.Part) (*parts.Part, int64, int64, error) {
+	return func(compact []*parts.Part) (*parts.Part, int64, int64, error) {
+		size, err := t.compactParts(writer, compact)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+
+		return nil, size, 0, nil
+	}
+}
+
+// compactParts will compact the given parts into a Parquet file written to w.
+// It returns the size in bytes of the compacted parts.
+func (t *Table) compactParts(w io.Writer, compact []*parts.Part) (int64, error) {
+	bufs := []dynparquet.DynamicRowGroup{}
+	var size int64
+	for _, part := range compact {
+		size += part.Size()
+		buf, err := part.AsSerializedBuffer(t.schema)
+		if err != nil {
+			return 0, err
+		}
+		bufs = append(bufs, buf.MultiDynamicRowGroup())
+	}
+	merged, err := t.schema.MergeDynamicRowGroups(bufs)
+	if err != nil {
+		return 0, err
+	}
+	err = func() error {
+		p, err := t.active.rowWriter(w, merged.DynamicColumns())
```

Comment:
Are we ok with not limiting the size in rows of the resulting compaction?

Reply:
The thing I'm most concerned about is that this will likely hurt the performance of queries with tight bounds given we apply the filter on a per-part basis. The larger the part, the less likely we can skip over the part (especially since parts that are a result of compaction are now overlapping, right?).

Reply:
Is that not where row groups would come into play? Since we filter at the row group level

Reply:
You might be right. I think it's worth running `BenchmarkQuery` to dispel any doubt on the perf of production queries.